ArcGISRuntimeSDKQt_CppSamples/Layers/ListKmlContents/ListKmlContents.cpp

```diff
@@ -0,0 +1,320 @@
+// [WriteFile Name=ListKmlContents, Category=Layers]
+// [Legal]
+// Copyright 2020 Esri.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// [Legal]
+
+#ifdef PCH_BUILD
+#include "pch.hpp"
+#endif // PCH_BUILD
+
+#include "ListKmlContents.h"
+
+#include "ArcGISTiledElevationSource.h"
+#include "KmlContainer.h"
+#include "KmlDataset.h"
+#include "KmlLayer.h"
+#include "KmlNetworkLink.h"
+#include "KmlNode.h"
+#include "KmlNodeListModel.h"
+#include "Scene.h"
+#include "SceneQuickView.h"
+
+#include <QDir>
+
+using namespace Esri::ArcGISRuntime;
+
+// helper method to get cross platform data path
+namespace
+{
+QString defaultDataPath()
+{
+  QString dataPath;
+
+#ifdef Q_OS_ANDROID
+  dataPath = "/sdcard";
+#elif defined Q_OS_IOS
+  dataPath = QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation);
+#else
+  dataPath = QDir::homePath();
+#endif
+
+  return dataPath;
+}
+} // namespace
+
+ListKmlContents::ListKmlContents(QObject* parent /* = nullptr */):
+  QObject(parent),
+  m_scene(new Scene(Basemap::imageryWithLabels(this), this))
+{
+  // create a new elevation source from Terrain3D REST service
+  ArcGISTiledElevationSource* elevationSource = new ArcGISTiledElevationSource(
+        QUrl("https://elevation3d.arcgis.com/arcgis/rest/services/WorldElevation3D/Terrain3D/ImageServer"), this);
+
+  // add the elevation source to the scene to display elevation
+  m_scene->baseSurface()->elevationSources()->append(elevationSource);
+
+  const QString filepath = defaultDataPath() + "/ArcGIS/Runtime/Data/kml/esri_test_data.kmz";
+
+  // create new KML layer
+  m_kmlDataset = new KmlDataset(filepath, this);
+  KmlLayer* kmlLayer = new KmlLayer(m_kmlDataset, this);
+  m_scene->operationalLayers()->append(kmlLayer);
+
+  connect(m_kmlDataset, &KmlDataset::doneLoading, this, [this](Error loadError)
+  {
+    if (!loadError.isEmpty())
+    {
+      qDebug() << loadError.message() << loadError.additionalMessage();
+      return;
+    }
+
+    // recursively build tree to display KML contents
+    for (int i = 0; i < m_kmlDataset->rootNodes().size(); ++i)
```

Comment:
can we do a range loop here?

```suggestion
    for (const KmlNode* node : qAsConst(m_kmlDataset->rootNodes()))
```